uniffi/src/lib.rs

```diff
@@ -223,24 +229,188 @@ unsafe impl ViaFfi for bool {
     }
 }
 
+/// Support for passing a buffer of bytes via the FFI.
+///
+/// We can pass a `Vec<u8>` to foreign language code by decomposing it into
+/// its raw parts (buffer pointer, length, and capacity) and passing those
+/// around as a struct. Naturally, this can be tremendously unsafe! So here
+/// are the details:
+///
+///   * `RustBuffer` structs must only ever be constructed from a `Vec<u8>`,
+///     either explicitly via `RustBuffer::from_vec` or indirectly by calling
+///     one of the `RustBuffer::new*` constructors.
+///
+///   * `RustBuffer` structs do not implement `Drop`, since they are intended
+///     to be passed to foreign-language code outside of the control of rust's
+///     ownership system. To avoid memory leaks they *must* passed back into
+///     rust and either explicitly destroyed using `RustBuffer::destroy`, or
+///     converted back to a `Vec<u8>` using `RustBuffer::destroy_into_vec`
+///     (which will then be dropped via rust's usual ownership-tracking system).
+///
+/// Implementation note: all the fields of this struct are private, so you can't
+/// manually construct instances that don't come from a `Vec<u8>`. If you've got
+/// a `RustBuffer` then it either came from a public constructor (all of which
+/// are safe) or it came from foreign-language code (which should be encforcing
+/// safety invariants).
+///
+/// This struct is based on `ByteBuffer` from the `ffi-support` crate, but modified
+/// to retain unallocated capacity rather than truncating to the occupied length.
+#[repr(C)]
+pub struct RustBuffer {
+    /// The allocated capacity of the underlying `Vec<u8>`.
+    /// In rust this is a `usize`, but we use an `i32` for compatibility with JNA.
+    capacity: i32,
```

Comment:
Note that this is different from `ffi_support`, which uses an `i64` for the `len` field. I think we're mostly sticking to `i32` for length-related things due to JVM limitations so I figured it was worth being consistent here.